215 loss parameters

acasamitjana opened the pull request:
# Description

I have added a kernel size and a kernel type option for LNCC image similarity loss. Hence, I've added the corresponding unit tests and updated the docstring.

Fixes #215 

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [X] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      and formatted all changed files.
- [X] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

acasamitjana commented:
Hey, I'd like to have a couple of reviewers. Do you have time earlier next week? It's easy @mathpluscode @s-sd @RemiDelaunay

mathpluscode commented on test/unit/test_loss_image.py:
```diff
@@ -31,6 +31,23 @@ def test_lncc(self):
         expected = [-1, -1]
         assert is_equal_tf(got, expected)
 
+        got = image.dissimilarity_fn(
```
emm sorry, seems like I didn't write the test very well the last time I changed it.

Could we use `parameterize` for the tests? I can do an example in `test_ssd(self)` ;)

s-sd reviewed (commented):
I don't have any additional concerns to what @mathpluscode has already highlighted. I think when those are resolved just mention or re-request a review from me and I'd be happy to take another look/ approve.

acasamitjana commented on deepreg/model/loss/image.py:
```diff
@@ -4,6 +4,7 @@
 import tensorflow as tf
 
 EPS = 1.0e-6  # epsilon to prevent NaN
+PI2_SQRT = 2.506628274631
```
it is sqrt(2·pi)

mathpluscode replied:
@acasamitjana would you like to put a comment to describe it?

or we can do like

```python
import numpy as np
SQRT_2PI = np.sqrt(2*np.pi)
```

this will be probably executed only once.

acasamitjana replied:
ok, I agree

mathpluscode commented:
@acasamitjana I've pushed a refactoring, where I moved the tests regarding the kernel type into a test dedicated to lncc, and used a class to group output test and error test.

Something interesting, or maybe a bug, we used to use [-1, -1] as expected values, but with this refactoring, the output should be [1, 1]. I'm not very familiar with this loss, does it make sense to you? seems like if y_pred = alpha * y_true, then the output is always 1...

```python
    y_true = tf.ones((2, 1, 2, 3, 2), dtype=tf.float32)
    y_pred = 0.5 * y_true

    @pytest.mark.parametrize(
        "y_true,y_pred,kernel_type,expected",
        [
            (y_true, y_pred, "rectangular", [1, 1]),
            (y_true, y_pred, "triangular", [1, 1]),
            (y_true, y_pred, "gaussian", [1, 1]),
            (y_pred, y_pred, "rectangular", [1, 1]),
            (y_pred, y_pred, "triangular", [1, 1]),
            (y_pred, y_pred, "gaussian", [1, 1]),
        ],
    )
```

acasamitjana commented:
Thanks @mathpluscode. I have refactored the remaining tests + correct a small bug in gmi and lncc functions. I think now it is ready to go. 

@s-sd we have updated the result, so it will be useful if you can take a look and approve/request changes. Let me know if you can't.

> Something interesting, or maybe a bug, we used to use [-1, -1] as expected values, but with this refactoring, the output should be [1, 1]. I'm not very familiar with this loss, does it make sense to you? seems like if y_pred = alpha * y_true, then the output is always 1...

That's right. We've made that even for negative alphas the result is 1 (convenient for the loss even though it is not the true value).

> btw, we are using `y_true = tf.constant(np.array(range(12)).reshape((2, 1, 2, 3)), dtype=tf.float32)` which has values >1,
> technically it's invalid data as we assume one-hot values.
> maybe it'd be better to divide it by 12 to at least limit the value between [0, 1]

Yes, if I remember correctly, we currently normalize images from [0,255] to [0,1]. It may be happening to other tests maybe, so either we open another ticket or we leave it (my option).

pD: I have removed the TODO line from GMI. I believe a certain tolerance may be allowed because we can't compute MI with infinite precision (we use a gaussian kernel to compute the pdf).

mathpluscode commented:
@acasamitjana we still need a small comment for the constant you defined ;) otherwise LGTM.

acasamitjana commented:
> @acasamitjana we still need a small comment for the constant you defined ;) otherwise LGTM.

Yep, sorry. I forgot to delete it. Since it's just a normalization constant, I delete it and kernel_vol variable normalizes the result later on.
